plugins/out_stackdriver/stackdriver.c

```diff
@@ -2534,36 +2538,43 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
     uint64_t ts = cfl_time_now();
 #endif
 
-    /* Get upstream connection */
-    u_conn = flb_upstream_conn_get(ctx->u);
-    if (!u_conn) {
+        /* Reformat msgpack to stackdriver JSON payload */
+    payload_buf = stackdriver_format(ctx,
+                                     event_chunk->total_events,
+                                     event_chunk->tag, flb_sds_len(event_chunk->tag),
+                                     event_chunk->data, event_chunk->size);
+    if (!payload_buf) {
 #ifdef FLB_HAVE_METRICS
         cmt_counter_inc(ctx->cmt_failed_requests,
                         ts, 1, (char *[]) {name});
 
         /* OLD api */
         flb_metrics_sum(FLB_STACKDRIVER_FAILED_REQUESTS, 1, ctx->ins->metrics);
-
-        update_http_metrics(ctx, event_chunk, ts, STACKDRIVER_NET_ERROR);
-        update_retry_metric(ctx, event_chunk, ts, STACKDRIVER_NET_ERROR, FLB_RETRY);
 #endif
+        flb_upstream_conn_release(u_conn);
         FLB_OUTPUT_RETURN(FLB_RETRY);
     }
 
-    /* Reformat msgpack to stackdriver JSON payload */
-    payload_buf = stackdriver_format(ctx,
-                                     event_chunk->total_events,
-                                     event_chunk->tag, flb_sds_len(event_chunk->tag),
-                                     event_chunk->data, event_chunk->size);
-    if (!payload_buf) {
+    if (ctx->test_log_entry_format) {
+        printf("%s", payload_buf);
+        flb_sds_destroy(payload_buf);
+        FLB_OUTPUT_RETURN(FLB_OK);
+        return;
+    }
+
+    /* Get upstream connection */
+    u_conn = flb_upstream_conn_get(ctx->u);
+    if (!u_conn) {
 #ifdef FLB_HAVE_METRICS
         cmt_counter_inc(ctx->cmt_failed_requests,
                         ts, 1, (char *[]) {name});
 
         /* OLD api */
         flb_metrics_sum(FLB_STACKDRIVER_FAILED_REQUESTS, 1, ctx->ins->metrics);
+
+        update_http_metrics(ctx, event_chunk, ts, STACKDRIVER_NET_ERROR);
+        update_retry_metric(ctx, event_chunk, ts, STACKDRIVER_NET_ERROR, FLB_RETRY);
 #endif
-        flb_upstream_conn_release(u_conn);
```

Comment:
You should now free the payload buffer here…

Reply:
This doesn't seem to be done. The original code first grabbed a connection, and then allocated the payload. If grabbing the connection failed, there was no payload to free yet, but when allocating the payload failed, the code had to release the connection. You changed the order of the two, so now if allocating the payload fails, there is no connection to release yet, but once it's allocated, failing to grab a connection means you have to free the payload (and, BTW, there's no connection to release in this branch).

To be clear, I mean this:
```suggestion
        flb_sds_destroy(payload_buf);
```